sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -184,7 +183,8 @@ class Dataset[T] private[sql](
       // to happen right away to let these side effects take place eagerly.
       case p if hasSideEffects(p) =>
         LogicalRDD(queryExecution.analyzed.output, queryExecution.toRdd)(sparkSession)
-      case Union(children) if children.forall(hasSideEffects) =>
+      case Union(children) if children.forall(x =>
+          x.isInstanceOf[InsertIntoTable] || x.isInstanceOf[InsertIntoHadoopFsRelationCommand]) =>
```

Comment:
Please consider the following example in the current master. This PR blocks this kind of weird behavior.
```
scala> spark.emptyDataFrame.createOrReplaceTempView("empty")
scala> sql("CREATE TABLE tbl(i INT) USING PARQUET")
scala> sql("(INSERT INTO tbl VALUES(1)) UNION ALL (INSERT INTO tbl VALUES(2))")
scala> sql("select * from tbl").collect()
res4: Array[org.apache.spark.sql.Row] = Array([1], [2])
scala> sql("(INSERT INTO tbl VALUES(5)) UNION ALL (SELECT * FROM empty)")
scala> sql("select * from tbl").collect()
res6: Array[org.apache.spark.sql.Row] = Array([1], [2])
```